### ISSUE TITLE
[Feature Request] random sort order

### ISSUE THREAD

**Issue by eminos** [typesense/typesense#637]
My use case for this is to fetch "a fixed number (6 for example) of popular products with these sets of attributes" to display in a grid/swiper layout on my ecommerce websites homepage. These need to be randomized.

I could fetch these products from the "originating" database, MySQL. But I'd rather just query the Typesense database, for several reasons (easy advanced filtering, etc.).

I would also like to "load more" random results and in that request add to exclude the already loaded products (IDs). I believe this part is already possible.

**Comment by jasonbosco** [comment]
One workaround for implementing this (until we implement this in Typesense) would be to add a new integer field called say `random_sort_field` to every record, and set it to a random number any time the record is created / updated in Typesense. 

Then at query time, you could run do a `filter_by` query like this: `filter_by: random_sort_field:=[1412..5922]` and also add a `sort_by: random_sort_field:desc` to sort randomly.

**Comment by casinojeffrey** [comment]
any update on implementing this in Typesense?

**Comment by JyTy** [comment]
Also interested on an update on when will the random sorting be supported out of the box?

**Comment by kishorenc** [comment]
@JyTy 

Is it enough if we added a boolean search parameter called `randomize` which will just re-order the results in the particular page requested randomly before being sent to the client? Also, I'm curious why this cannot be done on the client side: is it just more convenient to do it on server side?

**Comment by soukicz** [comment]
> Also, I'm curious why this cannot be done on the client side: is it just more convenient to do it on server side?

It is not the same because of pagination. If I want to get 5 random items I would have to search for 1000 items and then pick 5 random on client and it still isn't really random if whole dataset has milion items.
Closest solution is what @jasonbosco is suggesting but that does not work for small datasets. If I filter results to last 10 items and use random int, I will get no results. This is solvable on client side but only if I know in advance how many results will I get.

Really nice implementation is in [ElasticSearch](https://www.elastic.co/guide/en/elasticsearch/reference/current/query-dsl-function-score-query.html#function-random). It supports random sort but also seeded random - I can pass some string (for example IP address + current date) a item order will by random but stable for this string (I can paginate through this random list)

**Comment by kishorenc** [comment]
Thank you for the explanation @soukicz -- if you want the entire result set to be randomized, then I think we need to add a way to support that in the `sort_by` clause. Maybe something like:

```
"sort_by": "_rand(42)"
```

where `42` is a seed.

**Comment by JyTy** [comment]
Yep, I have a similar feature I need to deliver than @soukicz. We have a search built on top of TypeSense, and there is a feature request that when a user selects an entity, we should have a component on its page that shows four more random entities from the same category. The database is growing all the time so fetching thousands of entries just to pick 4 random ones does not seem effective.

Implementing some sort of random range on search time narrows the set a bit more, but it is hard to do in a way that all of entities would get same sort of chance for exposure... and this is really important here.

Previously we used an entirely different DB for that (Rockset - we needed it for other reasons but this was an added bonus) but they are shutting it down because they have sold to OpenAI.

**Comment by soukicz** [comment]
> Thank you for the explanation @soukicz -- if you want the entire result set to be randomized, then I think we need to add a way to support that in the `sort_by` clause. Maybe something like:
> 
> ```
> "sort_by": "_rand(42)"
> ```
> 
> where `42` is a seed.

Yes, that would be great! I have no idea how this could be implemented effectively to avoid generating random numbers for everything on each query.

_(BTW I am subscribed to this issue for two years because this one feature blocked our migration from elasticsearch two years ago :) )_

**Comment by ginorey** [comment]
Any update on this being added?

**Comment by kishorenc** [comment]
We do want to add this feature, it's on our medium term roadmap (next 4-6 weeks).

**Comment by servehub-llc** [comment]
@kishorenc  - Is this still on track based on your above comment?  I too urgently need this feature.

**Comment by kishorenc** [comment]
@servehub-llc 

Yes, will be able to share a build in a couple of days with this feature :)

**Comment by dfedak-agileti** [comment]
(+1) on this. We have a product index that seems to display default as "LIFO" on the product document insert / update.

**Comment by raphaelmprimo** [comment]
(+1) on this as well. Same scenario as other fellows, a random list of products for a marketplace.

I saw some commits coming about this feature request, very excited about it.

Are you guys planning to add this only on next full release (v28)? 

If yes, is that a way to get it before that, when you finish the development, like building on our own the source code from GitHub?

Thanks for this awesome open source search engine, btw.

**Comment by kishorenc** [comment]
Sorry that it took so long to get around to this, but I've just published `28.0.rc2` on which you can sort randomly like this:

```
"sort_by": "_rand:asc"
```

You can also pass a custom seed:

```
"sort_by": "_rand(113):asc"
```

Please try it out and let me know how it works.

**Comment by raphaelmprimo** [comment]
Thanks for the update @kishorenc 

Here is my feedback using the new `_rand` feature:
I tried with a 600k documents collection, using a search query + some filters + _rand and the CPU got a high usage suddenly

![Image](https://github.com/user-attachments/assets/0116a8cd-371a-4e0a-99f9-c8c2c43b1f36)

The request to the endpoint failed and only after few seconds the CPU got back to normal, and I was able to access again the search endpoint.

Let me know If you need complete details of the requests I tested to help.

**Comment by raphaelmprimo** [comment]
In the scenarios that the `_rand` function is working fine, the speed is up to **3x faster** than the custom random sort method that we implemented using data inside the document. Really good job! 👍🏼

**Comment by kishorenc** [comment]
@raphaelmprimo 

Can you please try again on `28.0.rc3`?

**Comment by raphaelmprimo** [comment]
@kishorenc 

Now it's working fast with `_text_match:desc,_rand:asc`!

I also noticed a faster search time with `_rand` in all previous scenarios with this new RC ⚡

**Comment by nbdesai1992** [comment]
So for the search results in question, would this _rand:asc field randomize all of the results in the collection? For example, what if I would like to randomize the first 20 results and then fall back to my ordering after that? I'm sorting my results based on an engagement score already, but want to add an element of randomness into the search if that makes sense so that users get some variety.

**Comment by sixers** [comment]
I think there's a bug - sorting with custom seed is not stable if query is `*` or empty string.

**Comment by kishorenc** [comment]
@sixers With wildcard searches, we parallelize some parts of the search operations, and that makes stability of the final results a bit difficult to pull off with randomization enabled. We are going to live with this limitation for now. Hopefully it's a big deal breaker.

**Comment by sixers** [comment]
It would be good to at least have an option to opt out of this optimisation and accept slower but stable searches.

**Comment by kishorenc** [comment]
I considered doing that, which is why I kept this PR open for a long time. Unfortunately, it's not easy to pull that off. Can you please create another issue for this: happy to take another look at our implementation to see if there is a work-around. I wanted to close this issue since we've mostly covered the underlying feature request, apart from this caveat.
